File: CPlusPlus/DesignPatterns/Singleton.cpp
// Header:
#include <iostream>

class Singleton
{
    /*
        Implementation notes:
            This is thread-safe as of C++ 11/17/..
            This is NOT thread-safe in C++ 98/03
    */

    public:
        // Public-facing methods:

        static void EnableLogging(const bool state)
        {
            Singleton& instance = GetInstance();
            instance.logEnabled = state;
        };

        static void LogMessage(const char* message)
        {
            Singleton& instance = GetInstance();
            if(instance.logEnabled)
            {
                std::cout << message << std::endl;
            }
        };

    private:
        // Private constructor:
        Singleton()
            : logEnabled(true)
        { };

        // Unlike other languages, we can declare it static
        // Makes instancing much nicer for us.
        static Singleton& GetInstance()
        {
            static Singleton instance;
            return instance;
        }

    public:
        // We don't want these to be usable:
        Singleton(const Singleton&) = delete;
        Singleton& operator=(const Singleton&) = delete;

    private:
        // Instance data:
        bool logEnabled;
};

File: TypeScript/DesignPatterns/Singleton.ts
// Singleton Demo in TypeScript

class Logger
{
    // Static instance:
    private static instance;

    // Instance data:
    private loggingEnabled: boolean;

    constructor()
    {
        this.loggingEnabled = true;
    }

    private static getInstance(): Logger
    {
        if(!Logger.instance)
        {
            Logger.instance = new Logger();
        }

        return Logger.instance;
    }

    // -------------------------------
    // Public-facing static methods:
    // -------------------------------
    
    public static enableLogging(state: boolean)
    {
        let logger = Logger.getInstance();
        logger.loggingEnabled = state;
    }

    public static log(message: any)
    {
        let logger = Logger.getInstance();
        if(logger.loggingEnabled)
        {
            console.log(message);
        }
    }
}